File: examples/php/error-handling.php
<?php

declare(strict_types=1);

// Typed error handling: branch on the specific exception subclass.
//   PUSHFY_API_TOKEN=... php error-handling.php
//
// Every failure throws a subclass of Pushfy\Exception\PushfyException, so a
// single catch can be a safety net while the specific ones drive behaviour.

require __DIR__ . '/vendor/autoload.php';

use Pushfy\Pushfy;
use Pushfy\Exception\AuthenticationException;
use Pushfy\Exception\RateLimitException;
use Pushfy\Exception\InvalidRequestException;
use Pushfy\Exception\ApiException;
use Pushfy\Exception\PushfyException;

$pushfy = new Pushfy([
    'apiToken' => getenv('PUSHFY_API_TOKEN') ?: '',
]);

try {
    $result = $pushfy->sms->send([
        'to'    => '5511999999999',
        'text'  => 'Testing typed error handling',
        'extId' => 'err-demo-' . time(),
    ]);
    echo "Accepted:\n";
    echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE), "\n";
} catch (RateLimitException $e) {
    // 429 — respect the server's backoff hint when present.
    $wait = $e->getRetryAfter() ?? 5;
    fwrite(STDERR, "Rate limited. Back off {$wait}s and retry.\n");
    exit(1);
} catch (AuthenticationException $e) {
    // 401/403 — bad or missing token/HMAC keys. Not retryable.
    fwrite(STDERR, "Auth failed (status {$e->getStatus()}). Check PUSHFY_API_TOKEN.\n");
    exit(1);
} catch (InvalidRequestException $e) {
    // 4xx — the request is wrong (bad phone, missing field). Fix and resend.
    fwrite(STDERR, sprintf(
        "Invalid request: status=%d code=%s\n",
        $e->getStatus(),
        (string) $e->getErrorCode()
    ));
    exit(1);
} catch (ApiException $e) {
    // 5xx / network / timeout — transient. Safe to retry with the SAME extId.
    fwrite(STDERR, sprintf(
        "Server/network error: status=%d code=%s — retry idempotently.\n",
        $e->getStatus(),
        (string) $e->getErrorCode()
    ));
    exit(1);
} catch (PushfyException $e) {
    // Catch-all safety net for anything not handled above.
    fwrite(STDERR, "Unexpected Pushfy error: {$e->getMessage()}\n");
    exit(1);
}

File: examples/python/error_handling.py
"""Branch on the typed errors the SDK raises.

Run:  PUSHFY_API_TOKEN=... python3 error_handling.py

Every failure raises a typed error you can catch. Order matters: catch the
specific subclasses before the ApiError base class.
"""

import os

from pushfy import (
    Pushfy,
    ApiError,
    AuthenticationError,
    InvalidRequestError,
    RateLimitError,
)


def main():
    pushfy = Pushfy(api_token=os.environ["PUSHFY_API_TOKEN"])

    try:
        result = pushfy.sms.send(
            to="5511999999999",
            text="Hello from Pushfy",
            ext_id="welcome-001",
        )
        print("Accepted:", result)

    except RateLimitError:
        # 429 — back off and retry later (see retry.py). Never tight-loop.
        print("Rate limited — back off and retry.")

    except AuthenticationError:
        # 401 — bad/missing token. Retrying won't help; fix the credentials.
        print("Authentication failed — check PUSHFY_API_TOKEN.")

    except InvalidRequestError as err:
        # 4xx — a bad parameter (invalid number, empty body, ...). Do not retry
        # as-is; fix the request first.
        print("Invalid request:", err.status, err.code, err.response)

    except ApiError as err:
        # 5xx / network / timeout — safe to retry idempotently by reusing the
        # SAME ext_id so a repeat can't create a duplicate charge.
        print("Server/network error:", err.status, err.code, err.response)


if __name__ == "__main__":
    main()
